subsys/usb/class/netusb/function_rndis.c

```diff
@@ -0,0 +1,1000 @@
+		return -EINVAL;
+	}
+
+	if (hdr->type != sys_cpu_to_le32(RNDIS_DATA_PACKET)) {
+		SYS_LOG_ERR("Wrong data packet type 0x%x",
+			    sys_le32_to_cpu(hdr->type));
+		return -EINVAL;
+	}
+
+	len = sys_le32_to_cpu(hdr->len);
+
+	if (len < sys_le32_to_cpu(hdr->payload_offset) +
+	    sys_le32_to_cpu(hdr->payload_len) +
+	    /*
+	     * Calculate additional offset since payload_offset is calculated
+	     * from the start of itself ;)
+	     */
+	    (u8_t *)&hdr->payload_offset - (u8_t *)hdr) {
+		SYS_LOG_ERR("Incorrect RNDIS packet");
+		return -EINVAL;
+	}
+
+	SYS_LOG_DBG("Parsing packet: len %u payload offset %u payload len %u",
+		    len, sys_le32_to_cpu(hdr->payload_offset),
+		    sys_le32_to_cpu(hdr->payload_len));
+
+	return len;
+}
+
+void rndis_clean(void)
+{
+	SYS_LOG_DBG("");
+
+	if (rndis.in_pkt) {
+		net_pkt_unref(rndis.in_pkt);
+
+		rndis.in_pkt = NULL;
+		rndis.in_pkt_len = 0;
+	}
+
+	rndis.skip_bytes = 0;
+}
+
+static void rndis_bulk_out(u8_t ep, enum usb_dc_ep_cb_status_code ep_status)
+{
+	u8_t buffer[CONFIG_RNDIS_BULK_EP_MPS];
+	u8_t hdr_offset = 0;
+	u32_t len;
+	int ret;
+
+	usb_read(ep, NULL, 0, &len);
+
+	SYS_LOG_DBG("EP 0x%x status %d len %u", ep, ep_status, len);
+
+	usb_read(ep, buffer, len, NULL);
+
+	/* We already keep frame keeping with len, warn here about
+	 * receiving frame delimeter
+	 */
+	if (len == 1 && !buffer[0]) {
+		SYS_LOG_DBG("Got frame delimeter, skip");
+		return;
+	}
+
+	/* Handle skip bytes */
+	if (rndis.skip_bytes) {
+		SYS_LOG_WRN("Skip %u bytes out of remaining %d bytes",
+			    len, rndis.skip_bytes);
+
+		rndis.skip_bytes -= len;
+
+		if (rndis.skip_bytes < 0) {
+			SYS_LOG_ERR("Error skipping bytes");
+
+			rndis.skip_bytes = 0;
+		}
+
+		return;
+	}
+
+	/* Start new packet */
+	if (!rndis.in_pkt) {
+		struct net_pkt *pkt;
+		struct net_buf *buf;
+
+		rndis.in_pkt_len = parse_rndis_header(buffer, len);
+		if (rndis.in_pkt_len < 0) {
+			SYS_LOG_ERR("Error parsing RNDIS header");
+
+			rndis.rx_err++;
+			return;
+		}
+
+		pkt = net_pkt_get_reserve_rx(0, K_NO_WAIT);
+		if (!pkt) {
+			/* In a case of low memory skip the whole packet
+			 * hoping to get buffers for later ones
+			 */
+			rndis.skip_bytes = rndis.in_pkt_len - len;
+			rndis.rx_no_buf++;
+
+			SYS_LOG_ERR("Not enough pkt buffers, len %u, skip %u",
+				    rndis.in_pkt_len, rndis.skip_bytes);
+
+			return;
+		}
+
+		buf = net_pkt_get_frag(pkt, K_NO_WAIT);
+		if (!buf) {
+			/* In a case of low memory skip the whole packet
+			 * hoping to get buffers for later ones
+			 */
+			rndis.skip_bytes = rndis.in_pkt_len - len;
+
+			SYS_LOG_ERR("Not enough net buffers, len %u, skip %u",
+				    rndis.in_pkt_len, rndis.skip_bytes);
+
+			net_pkt_unref(pkt);
+			rndis.rx_no_buf++;
+			return;
+		}
+
+		net_pkt_frag_insert(pkt, buf);
+
+		rndis.in_pkt = pkt;
+
+		/* Append data only, skipping RNDIS header */
+		hdr_offset = sizeof(struct rndis_payload_packet);
+	}
+
+	ret = net_pkt_append_all(rndis.in_pkt, len - hdr_offset,
```

Comment:
* `hdr_offset` should be declared as an `u32_t` to avoid converting it from `u8_t` here
* attributing to it the result of the `sizeof()` operator will truncate `size_t` to `u8_t` (as the code is written now); is this OK?
* Can `len` be smaller than `hdr_offset` in a way that the unsigned subtraction will lead to overflow?

Reply:
I have changed hdr_offset to u32_t though it cannot be so large since this is offset inside 64 bytes array.
hdr_offset is not zero if we get a new packet which starts with header of this offset size. In parse_rndis_header() we already have a check
	if (buf_len < sizeof(*hdr)) {
		SYS_LOG_ERR("Too small packet len %u", buf_len);
		return -EINVAL;
	}